source/robot_model/src/Model.cpp

```diff
@@ -29,17 +39,175 @@ bool Model::create_urdf_from_string(const std::string& urdf_string, const std::s
+        }
+        
+        start = matches_file[0].second;   
+    }
+    return extracted_words;
+}
+
+void Model::replace_package_by_full_path(const std::string& package_name, const std::string& full_path) {
+    std::string target = "package://" + package_name + "/";
+    std::string replacement = full_path;
+
+    std::string urdf = this->urdf_.str();
+
+    // Start from the beginning of the result string
+    size_t start_position = 0;
+    while ((start_position = urdf.find(target, start_position)) != std::string::npos) {
+        // Replace the target with the replacement string
+        urdf.replace(start_position, target.length(), replacement);
+        // Move past the last replacement
+        start_position += replacement.length();
+    }
+    this->urdf_.str(urdf);
+}
+
+void Model::resolve_package_paths() {
+    std::set<std::string> package_names = this->extract_package_name_from_urdf();
+
+    for (const auto& package_name : package_names) {
+        std::string package_path = this->meshloader_callback_(package_name);
+        this->package_paths_.push_back(package_path);
+        this->replace_package_by_full_path(package_name, package_path);
+    }
+}
+
+// read the URDF file and return the string
+std::stringstream Model::read_urdf_from_file() {
+    std::ifstream file_stream(this->get_urdf_path());
+    if (!file_stream.is_open()) {
+        throw std::runtime_error("Unable to open file: " + this->get_urdf_path());
+    }
+    std::stringstream buffer;
+    buffer << file_stream.rdbuf(); // Read the file content into the string stream
+
+    return buffer; // Convert the string stream to a string and return
+}
+
 void Model::init_model() {
-  pinocchio::urdf::buildModel(this->get_urdf_path(), this->robot_model_);
+  // initialize the collision package paths to empty
+  this->package_paths_ = {};
+
+  // check if the package path resolver callback is set then build model with the package resolver
+  if (this->load_collision_geometries_) { 
+    // read the URDF from file
+    this->urdf_ = this->read_urdf_from_file();
+
+    // resolve the package paths
+    this->resolve_package_paths();
+
+    // build the model with the resolved URDF
+    pinocchio::urdf::buildModelFromXML(this->urdf_.str(), this->robot_model_);
+
+    this->init_geom_model();
+  }
+  else {   // build the model without the package resolver
+    pinocchio::urdf::buildModel(this->get_urdf_path(), this->robot_model_);
+  }
+  
+  // robot data
   this->robot_data_ = pinocchio::Data(this->robot_model_);
+
   // get the frames
   std::vector<std::string> frames;
   for (auto& f : this->robot_model_.frames) {
     frames.push_back(f.name);
   }
   // remove universe and root_joint frame added by Pinocchio
   this->frames_ = std::vector<std::string>(frames.begin() + 2, frames.end());
-  this->init_qp_solver();
+  this->init_qp_solver();  
+}
+
+// Method to initialize collision geometries
+void Model::init_geom_model() {
```

Comment:
nitpick: what happens if `pinocchio::urdf::buildGeom` or other steps fail during this method? it might be nice to catch and rethrow a `robot_model::exceptions::CollisionGeometryException`.